File: Data Structures & Algorithms/anagram-groups/submission-4.py
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        
        buckets = dict()
        for string in strs:

            # normalize string
            # do a char count, format str like a1b0c2..z0
            count = [0] * 26
            res = ""
            for char in string:
                count[ord(char) - ord('a')] += 1
            
            for i in range(len(count)):
                res += str(i + ord('a')) + str(count[i])

            if res not in buckets:
                buckets[res] = [string]
            else:
                buckets[res].append(string)
        
        ret = []
        for key, value in buckets.items():
            ret.append(value)
        return ret

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-4.cpp
class Solution {
public:
    int characterReplacement(string s, int k) {
        vector<int> f(26);
        int l = 0, r = 0, maxf = 0;
        int res = 0;
        while (r < s.size()) {
            f[s[r] - 'A']++;
            maxf = max(maxf, f[s[r] - 'A']);

            while (r - l + 1 - maxf > k) {
                f[s[l] - 'A']--;
                l++;
            }

            res = max(res, r - l + 1);
            r++;
        }

        return res;
    }
};
